src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,216 @@
+#include <platform/ThreadStackManager.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+namespace ThreadBorderRouterManagement {
+
+class ScopedThreadLock
+{
+public:
+    ScopedThreadLock() { DeviceLayer::ThreadStackMgr().LockThreadStack(); }
+    ~ScopedThreadLock() { DeviceLayer::ThreadStackMgr().UnlockThreadStack(); }
+};
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::Init()
+{
+    mCallback = nullptr;
+    ReturnErrorOnFailure(DeviceLayer::PlatformMgrImpl().AddEventHandler(OnPlatformEventHandler, reinterpret_cast<intptr_t>(this)));
+    ReturnErrorOnFailure(RevertActiveDataset());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetBorderAgentId(MutableByteSpan & borderAgentIdSpan)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    otBorderAgentId borderAgentId;
+    if (borderAgentIdSpan.size() < sizeof(borderAgentId.mId))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    ScopedThreadLock threadLock;
+    otError err = otBorderAgentGetId(otInst, &borderAgentId);
+    if (err == OT_ERROR_NONE)
+    {
+        memcpy(borderAgentIdSpan.data(), borderAgentId.mId, sizeof(borderAgentId.mId));
+        borderAgentIdSpan.reduce_size(sizeof(borderAgentId.mId));
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetThreadVersion(uint16_t & threadVersion)
+{
+    threadVersion = otThreadGetVersion();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetInterfaceEnabled(bool & interfaceEnabled)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    ScopedThreadLock threadLock;
+    interfaceEnabled = otIp6IsEnabled(otInst);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetDataset(Thread::OperationalDataset & dataset, DatasetType type)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+
+    ScopedThreadLock threadLock;
+    otError otErr = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                                   uint32_t sequenceNum, ActivateDatasetCallback * callback)
+{
+    CHIP_ERROR err = BackupActiveDataset();
```

Comment:
In thread-border-router-management-server.cpp `HandleSetActiveDatasetRequest` method we have this code that follows the specification:
   
` // If this command is invoked when the ActiveDatasetTimestamp attribute is not null, the command SHALL
    // fail with a status code of INVALID_IN_STATE.
    if (mDelegate->GetDataset(currentActiveDataset, Delegate::DatasetType::kActive) == CHIP_NO_ERROR &&
        currentActiveDataset.GetActiveTimestamp(currentActiveDatasetTimestamp) == CHIP_NO_ERROR)
    {
        return Status::InvalidInState;
    }`

The OT delegate will never be invoked when there is an Active Dataset already configured so there will never be an Active Dataset to backup. Is my assumption correct? 
My understanding is that we can set the active dataset if none is configured, otherwise we need to use SetPendingDataset to update to a new one.

Reply:
+1, I think this makes sense

Reply:
Remove the backup active dataset function and added a function to store whether the Thread is commissioned.